internal/stats/apirequest.go

```diff
@@ -0,0 +1,95 @@
+package stats
+
+import (
+	"github.com/edulinq/autograder/internal/config"
+	"github.com/edulinq/autograder/internal/log"
+	"github.com/edulinq/autograder/internal/timestamp"
+)
+
+type APIRequestMetric struct {
+	BaseMetric
+
+	Sender       string `json:"sender"`
+	Endpoint     string `json:"endpoint"`
+	UserEmail    string `json:"user,omitempty"`
+	AssignmentID string `json:"assignment,omitempty"`
+	CourseID     string `json:"course,omitempty"`
+	Locator      string `json:"locator,omitempty"`
+
+	Duration uint64 `json:"duration"`
+}
+
+type APIRequestMetricQuery struct {
+	BaseQuery
+
+	Sender       string `json:"target-sender"`
+	Endpoint     string `json:"target-endpoint"`
+	UserEmail    string `json:"target-user,omitempty"`
+	CourseID     string `json:"target-course,omitempty"`
+	AssignmentID string `json:"target-assignment,omitempty"`
+	Locator      string `json:"target-locator"`
+}
+
+func (this APIRequestMetricQuery) Match(record *APIRequestMetric) bool {
+	if record == nil {
+		return false
+	}
+
+	if !this.BaseQuery.Match(record) {
+		return false
+	}
+
+	if (this.CourseID != "") && (this.CourseID != record.CourseID) {
+		return false
+	}
+
+	if (this.AssignmentID != "") && (this.AssignmentID != record.AssignmentID) {
+		return false
+	}
+
+	if (this.UserEmail != "") && (this.UserEmail != record.UserEmail) {
+		return false
+	}
+
+	if (this.Sender != "") && (this.Sender != record.Sender) {
+		return false
+	}
+
+	if (this.Endpoint != "") && (this.Endpoint != record.Endpoint) {
+		return false
+	}
+
+	if (this.Locator != "") && (this.Locator != record.Locator) {
+		return false
+	}
```

Comment:
I would change the order of these so that the most specific cases get checked first (this way we can best take advantage of short circuiting).